File: python/Trie.py
# ------------------------------------------------------------
# Data Structure: Trie (Prefix Tree)
# Difficulty: Medium
# Source: Standard Data Structure
# ------------------------------------------------------------
# Statement:
# Implement a Trie (Prefix Tree) with the following methods:
# - `insert(word)`: Inserts a word into the trie.
# - `search(word)`: Returns `True` if the word is in the trie.
# - `startsWith(prefix)`: Returns `True` if there is any word in the
#   trie that starts with the given prefix.
#
# A Trie is a tree-like data structure that stores a dynamic set of
# strings. It is commonly used for efficient retrieval of keys in a dataset
# of strings, such as autocomplete features or spell checkers.
#
# ------------------------------------------------------------

class TrieNode:
    """A node in the Trie structure."""
    def __init__(self):
        # `children` is a dictionary mapping a character to a TrieNode.
        self.children = {}
        # `is_end_of_word` is True if the node represents the end of a word.
        self.is_end_of_word = False

class Trie:
    """
    Trie (Prefix Tree) implementation.

    Approach:
    The Trie is built from a root node. Each node has a dictionary of children
    and a boolean flag to mark the end of a word.

    - `insert`: Traverse the trie character by character. If a character's
      node doesn't exist, create it. Mark the final node as the end of a word.
      Time Complexity: O(L), where L is the length of the word.
      Space Complexity: O(L) in the worst case (for a new word).

    - `search`: Traverse the trie. If any character is not found, the word
      doesn't exist. If the traversal completes, check if the final node
      is marked as the end of a word.
      Time Complexity: O(L).
      Space Complexity: O(1).

    - `startsWith`: Traverse the trie. If any character in the prefix is not
      found, no word starts with it. If traversal completes, the prefix exists.
      Time Complexity: O(P), where P is the length of the prefix.
      Space Complexity: O(1).
    """
    def __init__(self):
        """Initializes the trie structure."""
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """Inserts a word into the trie."""
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        """Returns True if the word is in the trie."""
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end_of_word

    def startsWith(self, prefix: str) -> bool:
        """Returns True if there is any word in the trie that starts with the given prefix."""
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True

# ------------------------------------------------------------
# Driver Code for Testing
# ------------------------------------------------------------
if __name__ == "__main__":
    trie = Trie()

    # --- Insert words ---
    words_to_insert = ["apple", "app", "apricot", "banana", "bandana"]
    for word in words_to_insert:
        trie.insert(word)
    print(f"Inserted words: {words_to_insert}\n")

    # --- Test search() method ---
    print("--- Testing search() ---")
    print(f"Search 'apple': {trie.search('apple')}")        # Expected: True
    print(f"Search 'app': {trie.search('app')}")          # Expected: True
    print(f"Search 'appl': {trie.search('appl')}")         # Expected: False
    print(f"Search 'banana': {trie.search('banana')}")      # Expected: True
    print(f"Search 'ban': {trie.search('ban')}")          # Expected: False
    print(f"Search 'bandanas': {trie.search('bandanas')}")  # Expected: False
    print("-" * 20)

    # --- Test startsWith() method ---
    print("--- Testing startsWith() ---")
    print(f"Starts with 'app': {trie.startsWith('app')}")      # Expected: True
    print(f"Starts with 'apri': {trie.startsWith('apri')}")   # Expected: True
    print(f"Starts with 'ban': {trie.startsWith('ban')}")      # Expected: True
    print(f"Starts with 'bana': {trie.startsWith('bana')}")    # Expected: True
    print(f"Starts with 'band': {trie.startsWith('band')}")    # Expected: True
    print(f"Starts with 'cat': {trie.startsWith('cat')}")      # Expected: False
    print(f"Starts with 'applepie': {trie.startsWith('applepie')}") # Expected: False
    print("-" * 20)

File: javascript/Trie.js
/**
 * Trie (Prefix Tree) Implementation in JavaScript
 * 
 * A Trie is a tree-like data structure that stores strings in a way that
 * allows for efficient prefix-based searching, autocomplete, and spell checking.
 * 
 * Time Complexity:
 * - Insert: O(m) where m is the length of the word
 * - Search: O(m) where m is the length of the word
 * - Prefix Search: O(m) where m is the length of the prefix
 * 
 * Space Complexity: O(n * m) where n is number of words and m is average length
 */

class TrieNode {
    constructor() {
        this.children = new Map(); // Stores character -> TrieNode mappings
        this.isEndOfWord = false;  // Marks if this node completes a word
        this.frequency = 0;        // Optional: track how many times word was inserted
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
        this.totalWords = 0;
    }

    /**
     * Insert a word into the trie
     * @param {string} word - The word to insert
     */
    insert(word) {
        if (!word || word.length === 0) return false;

        let currentNode = this.root;
        
        for (let char of word.toLowerCase()) {
            if (!currentNode.children.has(char)) {
                currentNode.children.set(char, new TrieNode());
            }
            currentNode = currentNode.children.get(char);
        }
        
        // Mark the end of word and update frequency
        if (!currentNode.isEndOfWord) {
            this.totalWords++;
        }
        currentNode.isEndOfWord = true;
        currentNode.frequency++;
        
        return true;
    }

    /**
     * Search for a complete word in the trie
     * @param {string} word - The word to search for
     * @returns {boolean} - True if word exists
     */
    search(word) {
        if (!word || word.length === 0) return false;

        let currentNode = this.root;
        
        for (let char of word.toLowerCase()) {
            if (!currentNode.children.has(char)) {
                return false;
            }
            currentNode = currentNode.children.get(char);
        }
        
        return currentNode.isEndOfWord;
    }

    /**
     * Check if any word in trie starts with given prefix
     * @param {string} prefix - The prefix to check
     * @returns {boolean} - True if prefix exists
     */
    startsWith(prefix) {
        if (!prefix || prefix.length === 0) return false;

        let currentNode = this.root;
        
        for (let char of prefix.toLowerCase()) {
            if (!currentNode.children.has(char)) {
                return false;
            }
            currentNode = currentNode.children.get(char);
        }
        
        return true;
    }

    /**
     * Get all words with given prefix (autocomplete)
     * @param {string} prefix - The prefix to autocomplete
     * @returns {string[]} - Array of matching words
     */
    autocomplete(prefix) {
        if (!prefix || prefix.length === 0) return [];
        
        let currentNode = this.root;
        const results = [];
        
        // Navigate to the prefix node
        for (let char of prefix.toLowerCase()) {
            if (!currentNode.children.has(char)) {
                return results; // No words with this prefix
            }
            currentNode = currentNode.children.get(char);
        }
        
        // Collect all words from this node
        this._collectWords(currentNode, prefix, results);
        return results;
    }

    /**
     * Helper method to collect all words from a node
     * @param {TrieNode} node - Current node
     * @param {string} currentWord - Current word being built
     * @param {string[]} results - Array to store results
     */
    _collectWords(node, currentWord, results) {
        if (node.isEndOfWord) {
            results.push({
                word: currentWord,
                frequency: node.frequency
            });
        }
        
        for (let [char, childNode] of node.children) {
            this._collectWords(childNode, currentWord + char, results);
        }
    }

    /**
     * Delete a word from the trie
     * @param {string} word - The word to delete
     * @returns {boolean} - True if word was deleted
     */
    delete(word) {
        if (!word || word.length === 0) return false;
        return this._deleteHelper(this.root, word.toLowerCase(), 0);
    }

    /**
     * Recursive helper for deletion
     */
    _deleteHelper(node, word, index) {
        if (index === word.length) {
            if (!node.isEndOfWord) return false;
            
            node.isEndOfWord = false;
            this.totalWords--;
            return node.children.size === 0; // Can be deleted if no children
        }
        
        const char = word[index];
        if (!node.children.has(char)) return false;
        
        const childNode = node.children.get(char);
        const shouldDeleteChild = this._deleteHelper(childNode, word, index + 1);
        
        if (shouldDeleteChild) {
            node.children.delete(char);
            return node.children.size === 0 && !node.isEndOfWord;
        }
        
        return false;
    }

    /**
     * Get the longest common prefix of all words
     * @returns {string} - Longest common prefix
     */
    longestCommonPrefix() {
        let prefix = "";
        let currentNode = this.root;
        
        while (currentNode.children.size === 1 && !currentNode.isEndOfWord) {
            const [char, childNode] = Array.from(currentNode.children)[0];
            prefix += char;
            currentNode = childNode;
        }
        
        return prefix;
    }

    /**
     * Get total number of words in trie
     * @returns {number} - Word count
     */
    getWordCount() {
        return this.totalWords;
    }

    /**
     * Get all words in trie
     * @returns {string[]} - All words
     */
    getAllWords() {
        const results = [];
        this._collectWords(this.root, "", results);
        return results.map(item => item.word).sort();
    }

    /**
     * Clear the entire trie
     */
    clear() {
        this.root = new TrieNode();
        this.totalWords = 0;
    }
}

// Example Usage and Test Cases
function demonstrateTrie() {
    console.log("=== Trie Data Structure Demonstration ===\n");
    
    const trie = new Trie();
    
    // Insert words
    const words = ["apple", "app", "application", "banana", "bat", "ball", "cat"];
    words.forEach(word => trie.insert(word));
    
    console.log("1. Inserted words:", words);
    console.log("Total words in trie:", trie.getWordCount());
    
    // Search operations
    console.log("\n2. Search Operations:");
    console.log("Search 'apple':", trie.search("apple"));     // true
    console.log("Search 'app':", trie.search("app"));         // true  
    console.log("Search 'appl':", trie.search("appl"));       // false
    console.log("Search 'dog':", trie.search("dog"));         // false
    
    // Prefix search
    console.log("\n3. Prefix Search:");
    console.log("StartsWith 'app':", trie.startsWith("app")); // true
    console.log("StartsWith 'ba':", trie.startsWith("ba"));   // true
    console.log("StartsWith 'dog':", trie.startsWith("dog")); // false
    
    // Autocomplete
    console.log("\n4. Autocomplete Suggestions:");
    console.log("Prefix 'app':", trie.autocomplete("app").map(item => item.word));
    console.log("Prefix 'ba':", trie.autocomplete("ba").map(item => item.word));
    console.log("Prefix 'c':", trie.autocomplete("c").map(item => item.word));
    
    // Delete operations
    console.log("\n5. Delete Operations:");
    console.log("Delete 'app':", trie.delete("app"));
    console.log("Search 'app' after deletion:", trie.search("app")); // false
    console.log("StartsWith 'app' after deletion:", trie.startsWith("app")); // true
    console.log("Autocomplete 'app' after deletion:", trie.autocomplete("app").map(item => item.word));
    
    // Longest common prefix
    console.log("\n6. Longest Common Prefix:", trie.longestCommonPrefix());
    
    // Get all words
    console.log("\n7. All words in trie:", trie.getAllWords());
}

// Advanced Example: Spell Checker using Trie
class SpellChecker {
    constructor() {
        this.trie = new Trie();
    }
    
    addToDictionary(words) {
        words.forEach(word => this.trie.insert(word));
    }
    
    suggestCorrections(word, maxSuggestions = 5) {
        if (this.trie.search(word)) {
            return [`"${word}" is spelled correctly!`];
        }
        
        const suggestions = this.trie.autocomplete(word.slice(0, 3));
        return suggestions
            .sort((a, b) => b.frequency - a.frequency)
            .slice(0, maxSuggestions)
            .map(item => item.word);
    }
}

// Demonstrate spell checker
function demonstrateSpellChecker() {
    console.log("\n\n=== Spell Checker Demonstration ===");
    
    const spellChecker = new SpellChecker();
    const dictionary = [
        "hello", "world", "javascript", "programming", 
        "algorithm", "data", "structure", "computer",
        "science", "code", "development", "web"
    ];
    
    spellChecker.addToDictionary(dictionary);
    
    console.log("Suggestions for 'prog':", spellChecker.suggestCorrections("prog"));
    console.log("Suggestions for 'dat':", spellChecker.suggestCorrections("dat"));
    console.log("Suggestions for 'xyz':", spellChecker.suggestCorrections("xyz"));
}

// Run demonstrations
if (require.main === module) {
    demonstrateTrie();
    demonstrateSpellChecker();
}

module.exports = { Trie, TrieNode, SpellChecker };